src/waveform/renderers/allshader/waveformrendererslipmode.cpp

```diff
@@ -1,116 +1,134 @@
-            m_waveformRenderer->getGroup(), "slip_enabled");
+    m_pSlipModeControl.reset(new ControlProxy(
+            m_waveformRenderer->getGroup(), "slip_enabled"));
 
     return true;
 }
 
-void WaveformRendererSlipMode::setup(const QDomNode& node, const SkinContext& context) {
-    const QString slipModeOutlineColorName = context.selectString(node, "SlipBorderOutlineColor");
+void WaveformRendererSlipMode::setup(const QDomNode& node, const SkinContext& skinContext) {
+    const QString slipModeOutlineColorName =
+            skinContext.selectString(node, "SlipBorderOutlineColor");
     if (!slipModeOutlineColorName.isNull()) {
         m_color = WSkinColor::getCorrectColor(QColor(slipModeOutlineColorName));
     } else {
         m_color = kDefaultColor;
     }
-    const float slipBorderTopOutlineSize = context.selectFloat(
+    const float slipBorderTopOutlineSize = skinContext.selectFloat(
             node, "SlipBorderTopOutlineSize", m_slipBorderTopOutlineSize);
     if (slipBorderTopOutlineSize >= 0) {
         m_slipBorderTopOutlineSize = slipBorderTopOutlineSize;
     }
-    const float slipBorderBottomOutlineSize = context.selectFloat(
+    const float slipBorderBottomOutlineSize = skinContext.selectFloat(
             node, "SlipBorderBottomOutlineSize", m_slipBorderBottomOutlineSize);
     if (slipBorderBottomOutlineSize >= 0) {
         m_slipBorderBottomOutlineSize = slipBorderBottomOutlineSize;
     }
 }
 
-void WaveformRendererSlipMode::initializeGL() {
-    m_shader.init();
+void WaveformRendererSlipMode::preprocess() {
+    if (!preprocessInner()) {
+        geometry().allocate(0);
+        markDirtyGeometry();
+    }
 }
 
-void WaveformRendererSlipMode::paintGL() {
-    if (!m_pSlipMode->toBool() || !m_waveformRenderer->isSlipActive()) {
-        return;
+bool WaveformRendererSlipMode::preprocessInner() {
+    if (!m_pSlipModeControl->toBool() || !m_waveformRenderer->isSlipActive()) {
+        return false;
     }
 
     const int elapsed = m_timer.elapsed().toIntegerMillis() % kBlinkingPeriodMillis;
 
-    const double blinkIntensity =
-            static_cast<double>(2 * abs(elapsed - kBlinkingPeriodMillis / 2)) /
+    const float blinkIntensity =
+            static_cast<float>(
+                    2 * std::abs(elapsed - kBlinkingPeriodMillis / 2)) /
             kBlinkingPeriodMillis;
 
-    const double alpha = 0.25 + 0.5 * blinkIntensity;
-
-    if (alpha != 0.0) {
-        QColor color = m_color;
-        color.setAlphaF(static_cast<float>(alpha));
-
-        glEnable(GL_BLEND);
-        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
-
-        const int colorLocation = m_shader.colorLocation();
-        const int borderLocation = m_shader.boarderLocation();
-        const int positionLocation = m_shader.positionLocation();
-        const int gradientLocation = m_shader.dimensionLocation();
+    const float alpha = std::clamp(0.25 + 0.5 * blinkIntensity, 0.0, 1.0);
```

Comment:
This is giving me a conversion warning here

```suggestion
    const float alpha = std::clamp(0.25f + 0.5f * blinkIntensity, 0.0f, 1.0f);
```